include/cuco/detail/static_map/static_map_ref.inl

```diff
@@ -585,14 +586,167 @@ class operator_impl<
+   * @tparam Init Type of init value convertible to payload type
+   * @tparam Op Callable type which is used as apply operation and can be
+   *   called with arguments as Op(cuda::atomic_ref<T, Scope>, T). Op strictly must
+   *   have this signature to atomically apply the operation.
+
+   * @param value The element to insert
+   * @param init The init value of the op
+   * @param op The callable object to perform binary operation between existing value at the slot
+   *  and the element to insert.
+   */
+  template <typename Value,
+            typename Init,
+            typename Op,
+            typename = cuda::std::enable_if_t<std::is_convertible_v<Value, value_type>>>
+  __device__ void insert_or_apply(Value const& value, Init init, Op op)
+  {
+    static_assert(cg_size == 1, "Non-CG operation is incompatible with the current probing scheme");
+
+    static_assert(
+      cuda::std::is_invocable_v<Op, cuda::atomic_ref<T, Scope>, T>,
+      "insert_or_apply expects `Op` to be a callable as `Op(cuda::atomic_ref<T, Scope>, T)`");
+
+    auto& ref_ = static_cast<ref_type&>(*this);
+    ref_.dispatch_insert_or_apply(value, init, op);
+  }
+
+  /**
+   * @brief Inserts a key-value pair `{k, v}` if it's not present in the map. Otherwise, applies
+   * `Op` binary function to the mapped_type corresponding to the key `k` and the value `v`.
+   *
+   * @tparam Value Input type which is implicitly convertible to 'value_type'
+   * @tparam Op Callable type which is used as apply operation and can be
+   *   called with arguments as Op(cuda::atomic_ref<T, Scope>, T). Op strictly must
+   *   have this signature to atomically apply the operation.
+   *
+   * @param group The Cooperative Group used to perform group insert
+   * @param value The element to insert
+   * @param op The callable object to perform binary operation between existing value at the slot
+   *  and the element to insert.
+   */
+
+  template <typename Value, typename Op>
+  __device__ void insert_or_apply(cooperative_groups::thread_block_tile<cg_size> const& group,
+                                  Value const& value,
+                                  Op op)
+  {
+    static_assert(
+      cuda::std::is_invocable_v<Op, cuda::atomic_ref<T, Scope>, T>,
+      "insert_or_apply expects `Op` to be a callable as `Op(cuda::atomic_ref<T, Scope>, T)`");
+    auto& ref_ = static_cast<ref_type&>(*this);
+
+    auto constexpr init_eq_sentinel = false;
+    ref_.insert_or_apply_impl<init_eq_sentinel>(group, value, op);
+  }
+
+  template <typename Value>
+  __device__ void insert_or_apply(cooperative_groups::thread_block_tile<cg_size> const& group,
+                                  Value const& value,
+                                  cuco::op::reduce::sum_tag)
+  {
+    auto& ref_ = static_cast<ref_type&>(*this);
+    ref_.insert_or_apply(
+      group, value, T{}, [](cuda::atomic_ref<T, Scope> payload_ref, T const& payload) {
```

Comment:
```suggestion
      group, value, [](cuda::atomic_ref<T, Scope> payload_ref, T const& payload) {
```
question: could we use the above overload (https://github.com/srinivasyadav18/cuCollections/blob/a30557d1c769b805ff664370794e03431c8bcccd/include/cuco/detail/static_map/static_map_ref.inl#L651-L663) to avoid runtime dispatching?

Reply:
Yes. Actually we can directly dispatch to `insert_or_apply_impl`.

Reply:
Sorry. No. We cannot do in that way. We have to dispatch to the `insert_or_apply` overload with `init` parameter, as we do not know if the map is intialized with init or not. So we still need to use `init` parameter to compare with the sentinel of map to dispatch.

Reply:
resolved in [720d448](https://github.com/NVIDIA/cuCollections/pull/555/commits/720d4482c3d33f95b96e44964184daa22e827b81)

Similar to **host-bulk** `insert_or_apply` with `op` and `init`, we need to another overload that takes `sum_tag` as `op`  with `init` parameter. This would enable the efficient code path when using `cuco::op::reduce*`.

Reply:
[fd68d04](https://github.com/NVIDIA/cuCollections/pull/555/commits/fd68d045a3c7e8e6b2f5a4e015aeff479ce18f28) replaces tag based approach with functors. This will elimante us maintaining the tags and init element. The users can now flexibly chose either of the overloads (**init** or **no init**) with `insert_or_apply`.